wgengine/magicsock/magicsock_linux.go

```diff
@@ -0,0 +1,316 @@
+// firewall; see issue 3824 for more detail.
+func (c *Conn) listenDisco() {
+	if debugDisableAF_PACKET {
+		return
+	}
+	c.listenDiscoProto(&c.bpfConn4, "IPv4", magicsockFilterV4)
+	c.listenDiscoProto(&c.bpfConn6, "IPv6", magicsockFilterV6)
+}
+
+func (c *Conn) listenDiscoProto(fdptr *int, name string, filter []bpf.Instruction) {
+	// We create our socket manually here, rather than using
+	// {ipv4,ipv6}.NewPacketConn, because they don't support SOCK_RAW.
+	fd, err := unix.Socket(unix.AF_PACKET, unix.SOCK_RAW, int(htons(unix.ETH_P_ALL)))
+	if err != nil {
+		c.logf("magicsock: error creating raw socket for %s: %v", name, err)
+		return
+	}
+
+	*fdptr = fd
+	go func() {
+		err := c.listenPacketsWithFilter(fd, filter, c.handleDiscoPacket)
+		if err != nil {
+			c.logf("magicsock: error listening for %s disco packets with AF_PACKET: %v", name, err)
+		}
+	}()
+}
+
+func (c *Conn) closeListenDisco() {
+	if c.bpfConn4 != 0 {
+		unix.Close(c.bpfConn4)
+	}
+	if c.bpfConn6 != 0 {
+		unix.Close(c.bpfConn6)
+	}
+}
+
+// handleDiscoPacket is called by listenPacketsWithFilter when a packet
+// matching a BPF filter is received. It can be passed either an IPv4 or IPv6
+// packet and will handle either case.
+func (c *Conn) handleDiscoPacket(packet []byte, from unix.Sockaddr) error {
+	pkt := gopacket.NewPacket(packet, layers.LayerTypeEthernet, gopacket.NoCopy)
+
+	var (
+		srcIP net.IP
+		is4   bool
+	)
+	if l := pkt.Layer(layers.LayerTypeIPv6); l != nil {
+		srcIP = l.(*layers.IPv6).SrcIP
+		is4 = false
+	} else if l := pkt.Layer(layers.LayerTypeIPv4); l != nil {
+		srcIP = l.(*layers.IPv4).SrcIP
+		is4 = true
+	} else {
+		// unknown type; non-fatal
+		return nil
+	}
+
+	udpL := pkt.Layer(layers.LayerTypeUDP)
+	if udpL == nil {
+		// unexpected, but non-fatal
+		return nil
+	}
+	udp := udpL.(*layers.UDP)
+	srcPort := udp.SrcPort
+
+	srcAddr, ok := netip.AddrFromSlice(srcIP)
+	if !ok {
+		return nil
+	}
+
+	src := netip.AddrPortFrom(srcAddr, uint16(srcPort))
+	success := c.handleDiscoMessage(udp.Payload, src, key.NodePublic{})
+	switch {
+	case success && is4:
+		metricRecvDiscoPacketIPv4.Add(1)
+	case !success && is4:
+		metricRecvDiscoPacketInvalidIPv4.Add(1)
+
+	case success && !is4:
+		metricRecvDiscoPacketIPv6.Add(1)
+	case !success && !is4:
+		metricRecvDiscoPacketInvalidIPv6.Add(1)
+	}
+
+	return nil
+}
+
+// listenPacketsWithFilter creates a new AF_PACKET socket, applies the given
+// BPF filter to it, and then calls 'cb' with all packets that are received
+// from the socket (and thus ones that match the given filter).
+func (c *Conn) listenPacketsWithFilter(fd int, filt []bpf.Instruction, cb func([]byte, unix.Sockaddr) error) error {
+	// This filter ignores all packets
+	zeroFilter := bpfFilter{bpf.RetConstant{Val: 0x0}}
+	if err := zeroFilter.ApplyTo(fd); err != nil {
+		return fmt.Errorf("applying zero BPF filter: %w", err)
+	}
+
+	// Drain any existing packets in the queue, since things can
```

Comment:
Unnecessary, the generic receive code will do the right thing even when handed non-disco packets, won't it?